File: MinimumWindowSubstring/MinimumWindowSubstring.cpp
/*
 * time: O(n), space: O(1)
 */
#include <iostream>
#include <string>
#include <map>
#include <climits>
using namespace std;

class Solution{
public:
  string minWindow(string S, string T){
    int minLen = INT_MAX;
    string minStr = "";

    int needFind[256] = {0};
    int hasFound[256] = {0};
    for(int i = 0; i < T.size(); i++)
      needFind[T[i]]++;

    int start = 0;
    int end = 0;
    int count = 0;
    while(end < S.size()){
      if(needFind[S[end]] == 0){
        end++;
        continue;
      }

      if(hasFound[S[end]] < needFind[S[end]])
        count++;
      hasFound[S[end]]++;

      if(count < T.size()){
        end++;
        continue;
      }

      while(needFind[S[start]] == 0 || hasFound[S[start]] > needFind[S[start]]){
        if(hasFound[S[start]] > needFind[S[start]])
          hasFound[S[start]]--;
        start++;
      }
      int len = end - start + 1;
      if(len < minLen){
        minLen = len;
        minStr = S.substr(start, minLen);
      }

      end++;
    }

    return minStr;
  }
};


class Solution1{
public:
  string minWindow(string S, string T){
    string minStr = "";
    int minLen = INT_MAX;

    int needFind[256] = {0};
    int hasFound[256] = {0};
    for(int i = 0; i < T.size(); i++)
      needFind[T[i]]++;

    int start = 0;
    int end = 0;
    int count = 0;
    while(end < S.size()){
      if(hasFound[S[end]] < needFind[S[end]])
        count++;
      hasFound[S[end]]++;

      if(count < T.size()){
        end++;
        continue;
      }

      while(hasFound[S[start]] > needFind[S[start]]){
        hasFound[S[start]]--;
        start++;
      }
      int len = end - start + 1;
      if(len < minLen){
        minLen = len;
        minStr = S.substr(start, len);
      }
      end++;
    }
    return minStr;
  }
};

int main(){
  string S = "ADOBECODEBANC";
  string T = "ABC";

  Solution solution;
  string res = solution.minWindow(S, T);
  cout << res << endl;

  return 0;
}

File: MinimumWindowSubstring/MinimumWindowSubstring.py
class Solution(object):
    def minWindow(self, s, t):
        from collections import defaultdict
        import sys

        # save the freq of each char in t, it is a histogram
        d = defaultdict(int)
        for c in t:
            d[c] += 1

        # the number of chars need to find
        count = len(d)

        start, end = 0, 0
        min_len = sys.maxint
        min_str = ""

        while end < len(s):
            c = s[end]
            if c in d:
                d[c] -= 1
                if d[c] == 0:
                    count -=1
            end += 1

            # nothing to find because we've found everything needed
            while count == 0:
                tmpc = s[start]
                if tmpc in d:
                    d[tmpc] += 1
                    if d[tmpc] > 0:
                        count += 1
                if end - start < min_len:
                    min_len = end - start
                    min_str = s[start:end]
                start +=1

        return min_str
